builtin/logical/mongodb/path_roles.go

```diff
@@ -0,0 +1,228 @@
+	if roleDB == "" {
+		return logical.ErrorResponse("db parameter is required"), nil
+	}
+
+	// Example roles JSON:
+	//
+	// [ "readWrite", { "role": "readWrite", "db": "test" } ]
+	//
+	// For storage, we convert such an array into a homogeneous array of role documents like:
+	//
+	// [ { "role": "readWrite" }, { "role": "readWrite", "db": "test" } ]
+	//
+	var roles []mongodbRole
+	rolesJson := []byte(data.Get("roles").(string))
+	if len(rolesJson) > 0 {
+		var rolesArray []interface{}
+		err := json.Unmarshal(rolesJson, &rolesArray)
+		if err != nil {
+			return nil, err
+		}
+		for _, rawRole := range rolesArray {
+			switch role := rawRole.(type) {
+			case string:
+				roles = append(roles, mongodbRole{Role: role})
+			case map[string]interface{}:
+				if db, ok := role["db"].(string); ok {
+					if roleName, ok := role["role"].(string); ok {
+						roles = append(roles, mongodbRole{Role: roleName, DB: db})
+					}
+				}
+			}
+		}
+	}
+
+	// Store it
+	entry, err := logical.StorageEntryJSON("role/"+name, &roleStorageEntry{
+		DB:           roleDB,
+		MongoDBRoles: roles,
+	})
+	if err != nil {
+		return nil, err
+	}
+	if err := req.Storage.Put(entry); err != nil {
+		return nil, err
+	}
+
+	return nil, nil
+}
+
+func (roles mongodbRoles) toStandardRolesArray() []interface{} {
+	// Convert array of role documents like:
+	//
+	// [ { "role": "readWrite" }, { "role": "readWrite", "db": "test" } ]
+	//
+	// into a "standard" MongoDB roles array containing both strings and role documents:
+	//
+	// [ "readWrite", { "role": "readWrite", "db": "test" } ]
+	//
+	// MongoDB's createUser command accepts the latter.
+	//
+	var standardRolesArray []interface{}
+	for _, role := range roles {
+		if role.DB == "" {
+			standardRolesArray = append(standardRolesArray, role.Role)
+		} else {
+			standardRolesArray = append(standardRolesArray, role)
+		}
+	}
+	return standardRolesArray
+}
+
+type roleStorageEntry struct {
+	DB           string       `json:"db"`
+	MongoDBRoles mongodbRoles `json:"roles"`
+}
+
+type mongodbRole struct {
+	Role string `json:"role" bson:"role"`
+	DB   string `json:"db"   bson:"db"`
+}
+
+type mongodbRoles []mongodbRole
+
+const pathRoleHelpSyn = `
+Manage the roles used to generate MongoDB credentials.
+`
+
+const pathRoleHelpDesc = `
+This path lets you manage the roles used to generate MongoDB credentials.
+
+The "db" parameter specifies the authentication database for users
+generated for a given role.
+
+The "roles" parameter specifies the MongoDB roles that should be assigned
+to users created for a given role. Just like when creating a user directly
+using db.createUser, the roles JSON array can specify both built-in roles
+and user-defined roles for both the database the user is created in and
+for other databases.
+
+For example, the following roles JSON array grants the "readWrite"
+permission on both the user's authentication database and the "test"
+database:
+
+[ "readWrite", { "role": "readWrite", "db": "test" } ]
```

Comment:
Sorry, I'm not really following. It is valid to use that roles JSON array in a `vault write`:

```
vault write mongodb/roles/readonly roles='[ "readWrite", { "role": "read", "db": "bar" } ]'
```

Such a roles array is "MongoDB-standard" as written. It will be stored a bit differently by Vault, but when reading the role it will come back out in that same format...